Restore batch processing loop

thenbrent opened the pull request:
This PR reintroduces a loop in `ActionScheduler_QueueRunner::run()`  to increase the rate of action processing. Instead of processing just one batch of actions, with this patch, AS will process as many batches as it can before reaching 90% of the available memory, or it becomes likely processing another action will exceed the available script timeout limit.

Fixes #184

I recommend reviewing individual commit messages for a more in-depth explanation on each change.

---

I also snuck in https://github.com/Prospress/action-scheduler/commit/1a698d04795bac737dfa54769804b5db15c33fd8 to this PR, as it's a relatively small patch, which fixes #188, and it depends on the new `ActionScheduler_Compatibility` class being added here for #184, so I figured it was worth including in the same PR.

I'll move it to a separate PR if it slows down the review/merge of the patches required for #184, as that is a higher priority issue.

JPry commented on classes/ActionScheduler_Abstract_QueueRunner.php:
```diff
@@ -79,6 +92,115 @@ public function get_allowed_concurrent_batches() {
+		$max_execution_time  = $this->get_maximum_execution_time();
+		$time_per_action     = $execution_time / $processed_actions;
+		$estimated_time      = $execution_time + ( $time_per_action * 2 );
+		$likely_to_be_exceed = ( $estimated_time > $this->get_maximum_execution_time() ) ? true : false;
```
You can drop `? true : false` from the end of this line. This is redundant since the `>` operator (and other comparison operators) will return a boolean value. You can also omit the parentheses around the condition.

One minor thing here is the name of the variable. I'd suggest either `$likely_to_exceed` or `$likely_to_be_exceeded`.

JPry commented on classes/ActionScheduler_Abstract_QueueRunner.php:
```diff
@@ -79,6 +92,115 @@ public function get_allowed_concurrent_batches() {
+
+		$memory_limit    = $this->get_memory_limit() * 0.90;
+		$current_memory  = memory_get_usage( true );
+		$memory_exceeded = ( $current_memory >= $memory_limit ) ? true : false;
```
This is another place where you can drop the `? true : false` portion, as well as the parentheses.

JPry commented on classes/ActionScheduler_Abstract_QueueRunner.php:
```diff
@@ -22,6 +32,9 @@
 	public function __construct( ActionScheduler_Store $store = null, ActionScheduler_FatalErrorMonitor $monitor = null, ActionScheduler_QueueCleaner $cleaner = null ) {
+
+		$this->created_time = microtime( true );
+
```
Minor: We're just setting class properties here, so the extra new lines before and after this new statement can be omitted.

thenbrent replied:
@JPry then I'd have to make sure the `=` all align, which makes the diff harder to read. This property also serves a different purpose to the others, so I think it's justified it stands alone.

JPry commented on classes/ActionScheduler_Compatibility.php:
```diff
@@ -0,0 +1,79 @@
+	public static function raise_memory_limit() {
+		if ( function_exists( 'wp_raise_memory_limit' ) ) {
+			return wp_raise_memory_limit( 'admin' );
+		} else {
```
Putting all of this logic in an `else` block isn't necessary, since the function will return early if the WP function exists. It should be removed from the `else` block to help with readability by reducing indentation.

JPry commented on classes/ActionScheduler_Compatibility.php:
```diff
@@ -0,0 +1,79 @@
+	public static function convert_hr_to_bytes( $value ) {
+		if ( function_exists( 'wp_convert_hr_to_bytes' ) ) {
+			return wp_convert_hr_to_bytes( $value );
+		} else {
```
Similar to what was mentioned below, enclosing this section in an `else` block is unnecessary.

JPry commented on classes/ActionScheduler_Compatibility.php:
```diff
@@ -0,0 +1,79 @@
+					return false;
+				}
+			}
+		}
```
I don't *think* it's possible to get through this block without hitting one of the `return` conditions above, but just to be on the safe side I think we should include a `return false` at the very end of this function as a catch-all.

thenbrent commented:
Thanks for the review @JPry. I've addressed all of those issues and rebased in the fixes.

> I'm not opposed to this in principle. In relation to the discussion on Prospress/woocommerce-subscriptions#2851, would that change be better placed in a separate helper library?

Yes definitely, but we need this now in AS, so I don't the creation of the library to become a blocker for this to be added. We can migrate to use the library later once it's available.
